documents/dev_designs/tachyon_system.md

```diff
@@ -0,0 +1,220 @@
+* Support adding and removing nodes to the cluster to adapt to load.
+* Support player reconnection. A player that loses connection and reconnect
+  within a small window of time shouldn't be impacted: keep its place in any
+  matchmaking queue or lobby.
+* Support seamless server restart: lobbies, matchmaking queues, chat and so on
+  shouldn't be impacted when a server (maybe the only one) restarts.
+  This is only for graceful shutdown of nodes. Supporting these features
+  for sudden crash is a stretch goal.
+
+Tachyon is a hard requirement for the scaling goal. The other could be done
+with existing spring lobby protocol, but supporting tachyon requires significant
+changes to existing components so we might as well aim for all of them.
+
+## Player and connection
+
+A player connects to teiserver following the [tachyon doc](https://github.com/beyond-all-reason/tachyon/blob/master/docs/connection.md).
+
+Entities:
+
+* `PlayerConnection`: this is the process handling the http connection then upgraded
+  to websocket. It holds the transient state of a player like lobby membership.
+* `PlayerSessionSupervisor`: supervisor for `PlayerSession`
+* `PlayerSession`: A separate process, unique per player across the cluster.
+  Used to track if a given player is actually connected. Shouldn't hold much state.
+* `PlayerRegistry`: Global registry for players. Similar to the existing
+  `Teiserver.ClientRegistry` in purpose. May be able to reuse that directly.
+
+Example of an interaction where the same player connects, then connects again.
+
+```mermaid
+
+sequenceDiagram
+    participant Connection2
+    participant Connection
+    participant Registry
+    participant Session
+    participant SessionSupervisor
+    Connection ->> Registry: register
+    Registry ->> Connection: {:ok, <pid>}
+
+    Connection ->>+ SessionSupervisor: start session
+    SessionSupervisor ->> Session: start
+    SessionSupervisor ->> -Connection: {:ok, <pid>}
+    Session ->> Connection: monitor
+
+    Connection2 ->> Registry: register
+    Registry ->> Connection2: {:error, {:already_started, <pid>}}
+
+    Connection2 ->> +Session: update player
+    Session ->> Connection2: monitor
+    Session ->> Connection: force disconnect + demonitor
+    Session ->> -Connection2: :ok
+
+    Connection2 ->> Connection2: disconnect
+    Connection2 ->> Session: {:DOWN, _, :process, _, _}
+    Session ->> Session: wait until timeout
+```
+
+## Autohosts
+
+Autohosts connects to the server and are ready to be summoned to host a game by a lobby.
+Maybe expand tachyon so that they can provide some info about themselves, like region, engine version?
+
+Later, we can expand the autohost system to have some form of capacity scheduling.
+
+## Chatrooms
+Separate processes dedicated to relay messages across a list of recipients (player or autohost).
+Each process has a ring buffer of recent messages to allow replaying messages in case a player
+reconnects after crash.
```

Comment:
I assume chatrooms are a very generic thing that can be associated with lobby but maybe also for other purposes. I think the interesting part in that is if we ever wanted to support anything like chatroom history, e.g. modeling moderaion messages as DMs via special channel? Also, do people see room history from before they joined?

I'm not sure about any of that.

Reply:
You're correct, the chatroom bit isn't attached to lobby. I was thinking of the ability to attach one to a matchmaking queue for example.
Yes, chatroom will have the responsibility to persist durably the messages (for moderation purposes), I'm going to use the same existing code for that.

As for the history, maybe. I was thinking of having only the recent history in memory that can be retrieved, to support a player crashing and reconnecting so they can catch up. Not sure if a new player should be able to see chat history prior to joining, I would err on the side of "no" for that one, but not sure.